concent_api/core/models.py

```diff
@@ -28,6 +33,56 @@
 from .constants             import MESSAGE_TASK_ID_MAX_LENGTH
 
 
+class SubtaskWithTimingColumnsManager(Manager):
+    """Creates maximum_download time, subtask_verification_time and download_deadline columns
+    maximum_download_time = DOWNLOAD_LEADIN_TIME + ceil((result_package_size / MINIMUM_UPLOAD_RATE << 10))
+    subtask_verification_time = (4 * CONCENT_MESSAGING_TIME) + (3 * maximum_download_time) + (0.5 * (computation_deadline - task_to_compute_timestamp))
+    download_deadline = computation_deadline + maximum_download_time
+    """
+
+    @classmethod
+    def with_maximum_download_time(cls, query_set):
+        bytes_per_sec = settings.CONCENT_UPLOAD_RATE << 10
+        download_time = Func(F('result_package_size') / float(bytes_per_sec), function='CEIL')
+        return query_set.annotate(
+            maximum_download_time=Value(settings.DOWNLOAD_LEADIN_TIME) + download_time
+        )
+
+    @classmethod
+    def with_subtask_verification_time(cls, query_set):
+        computation_deadline = F('computation_deadline')
+        task_to_compute_timestamp = Func(Value('epoch'), F('task_to_compute__timestamp'), function='DATE_PART')
+        subtask_timeout = Func(Value('epoch'), computation_deadline, function='DATE_PART') - task_to_compute_timestamp
+        subtask_verification_time = (
+            4 * settings.CONCENT_MESSAGING_TIME) + (
+            3 * F('maximum_download_time')) + (
+            0.5 * subtask_timeout
+        )
+        return query_set.annotate(
+            subtask_verification_time=ExpressionWrapper(subtask_verification_time, output_field=IntegerField())
+        )
+
+    @classmethod
+    def with_download_deadline(cls, query_set):
+        computation_deadline = F('computation_deadline')
```

Comment:
This variable does not add much value. It would be better to put `F('computation_deadline')` directly in the query.

Reply:
Done.